### PR TITLE
Optionally execute range and regex terms as delayed iterators

### PR DESCRIPTION
**Pull request by apmoriarty**
This unit of work allows range and regex terms to execute as context-required (delayed) iterators instead of ivarators when the overall cardinality of a query is below a configured threshold.

This addresses the case where a low cardinality term is intersected with a regex term that has a high cardinality search space.

### CODE REVIEW

File: warehouse/query-core/src/main/java/datawave/query/iterator/QueryOptions.java [hunk 1]
```diff
@@ -1501,6 +1503,16 @@ public boolean validateOptions(Map<String,String> options) {
             this.termCounts = getMapSerDe().deserializeFromString(serializedMap);
         }
 
+        //
```
Comment (foster33) [inline warehouse/query-core/src/main/java/datawave/query/iterator/QueryOptions.java]: empty comment

File: warehouse/query-core/src/main/java/datawave/query/ancestor/AncestorQueryIterator.java [hunk 2]
```diff
@@ -128,10 +128,10 @@ public void seek(Range range, Collection<ByteSequence> columnFamilies, boolean i
 
     @Override
     public EventDataQueryFilter getEvaluationFilter() {
-        if (evaluationFilter == null && script != null) {
+        if (evaluationFilter == null && getScript() != null) {
 
             AttributeFactory attributeFactory = new AttributeFactory(typeMetadata);
-            Map<String,ExpressionFilter> expressionFilters = getExpressionFilters(script, attributeFactory);
+            Map<String,ExpressionFilter> expressionFilters = getExpressionFilters(getScript(), attributeFactory);
```
Comment (alerman) [inline warehouse/query-core/src/main/java/datawave/query/ancestor/AncestorQueryIterator.java:134]: Is this at all related to what you are changing, or just cleanup?

**Review by alerman — DISMISSED** [review]

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/CardinalityVisitor.java [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+import org.apache.commons.jexl3.parser.ASTReferenceExpression;
+import org.apache.commons.jexl3.parser.JexlNode;
+
+import datawave.query.jexl.nodes.QueryPropertyMarker;
+import datawave.query.jexl.nodes.QueryPropertyMarker.Instance;
+import datawave.query.jexl.nodes.QueryPropertyMarker.MarkerType;
+import datawave.query.util.count.CountMap;
+
+/**
+ * Finds the maximum cardinality of a query in the context of the field index
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/CardinalityVisitor.java]: Please define cardinality here

File: warehouse/query-core/src/main/java/datawave/query/iterator/QueryOptions.java [hunk 4]
```diff
@@ -2362,4 +2375,24 @@ public Equality getEquality() {
     public boolean isSeekingEventAggregation() {
         return seekingEventAggregation;
     }
+
+    public ASTJexlScript getScript() {
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/iterator/QueryOptions.java:2383]: Thank you for refactoring this here....

File: warehouse/query-core/src/main/java/datawave/query/iterator/logic/RangeFilterIterator.java [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+import datawave.query.Constants;
+import datawave.query.attributes.AttributeFactory;
+import datawave.query.attributes.Document;
+import datawave.query.data.parsers.FieldIndexKey;
+import datawave.query.iterator.NestedIterator;
+import datawave.query.iterator.SeekableIterator;
+import datawave.query.jexl.DelayedNonEventIndexContext;
+
+/**
+ * Allows a range term to be run as a context iterator
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/iterator/logic/RangeFilterIterator.java:29]: The term "context iterator" is not necessarily well known.  A fuller description of what this iterator does and in what context it is used would be appreciated.  This applies to the RegexFilterIterator as well.

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/CardinalityVisitor.java [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+    }
+
+    @Override
+    public Object visit(ASTOrNode node, Object data) {
+        long count = 0L;
+        for (int i = 0; i < node.jjtGetNumChildren(); i++) {
+            JexlNode child = node.jjtGetChild(i);
+            long childCount = (long) child.jjtAccept(this, data);
+
+            if (count == 0L) {
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/CardinalityVisitor.java]: We can break out of this loop once count is Long.MAX_VALUE.

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/CardinalityVisitor.java [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+    }
+
+    @Override
+    public Object visit(ASTAndNode node, Object data) {
+
+        Instance instance = QueryPropertyMarker.findInstance(node);
+        if (instance.isAnyTypeOf(MarkerType.EXCEEDED_TERM, MarkerType.EXCEEDED_VALUE, MarkerType.EXCEEDED_OR, MarkerType.BOUNDED_RANGE)) {
+            return Long.MAX_VALUE;
+        } else if (instance.isAnyTypeOf(Set.of(MarkerType.DELAYED, MarkerType.EVALUATION_ONLY))) {
+            return 0L;
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/CardinalityVisitor.java]: This is probably not needed given the follow on case

**Review by ivakegg — CHANGES_REQUESTED** [review]
a few optimizations, and definitly fuller documentation required please.

**Review by alerman — DISMISSED** [review]

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/CardinalityVisitor.java [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+
+    @Override
+    public Object visit(ASTOrNode node, Object data) {
+        long count = 0L;
+        for (int i = 0; i < node.jjtGetNumChildren(); i++) {
+            JexlNode child = node.jjtGetChild(i);
+            long childCount = (long) child.jjtAccept(this, data);
+
+            if (count == 0L) {
+                count = childCount;
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/CardinalityVisitor.java]: Could childCount be Long.MAX_VALUE here in which case we need to break out of the loop?

Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/jexl/visitors/CardinalityVisitor.java]: Yes. Took a pass through here and simplified the logic, also accounted for a long overflow condition.

**Review by ivakegg — DISMISSED** [review]

**Review by ivakegg — APPROVED** [review]

**Review by alerman — APPROVED** [review]

**Pull request merged**